engine/table/src/main/java/io/deephaven/engine/table/impl/by/AggregationRowLookup.java

```diff
@@ -0,0 +1,54 @@
+package io.deephaven.engine.table.impl.by;
+
+import static io.deephaven.engine.rowset.RowSequence.NULL_ROW_KEY;
+
+/**
+ * Tool to identify the aggregation result row key (also row position) from a logical key representing a set of values
+ * for the aggregation's group-by columns.
+ */
+public interface AggregationRowLookup {
+
+    /**
+     * Re-usable empty key, for use in (trivial) reverse lookups against no-key aggregations.
+     */
+    Object[] EMPTY_KEY = new Object[0];
+
+    /**
+     * Re-usable unknown row constant to serve as the default return value for {@link #noEntryValue()}.
+     */
+    int DEFAULT_UNKNOWN_ROW = (int) NULL_ROW_KEY;
+
+    /**
+     * Gets the row key value where {@code key} exists in the aggregation result table, or the {@link #noEntryValue()}
+     * if {@code key} is not found in the table.
+     * <p>
+     * This serves to map group-by column values to the row position (also row key) in the result table. Missing keys
+     * will map to {@link #noEntryValue()}.
+     * <p>
+     * Keys are specified as follows:
+     * <dl>
+     * <dt>No group-by columns</dt>
+     * <dd>"Empty" keys are signified by the {@link AggregationRowLookup#EMPTY_KEY} object, or any zero-length
+     * {@code Object[]}</dd>
+     * <dt>One group-by column</dt>
+     * <dd>Singular keys are (boxed, if needed) objects</dd>
+     * <dt>Multiple group-by columns</dt>
+     * <dd>Compound keys are {@code Object[]} of (boxed, if needed) objects, in the order of the aggregation's group-by
+     * columns</dd>
+     * </p>
+     * <p>
+     * All key fields must be reinterpreted to the appropriate primitive value before boxing. See
+     * {@link io.deephaven.engine.table.impl.sources.ReinterpretUtils#maybeConvertToPrimitive}.
+     *
+     * @param key A single (boxed) value for single-column keys, or an array of (boxed) values for compound keys
+     * @return The row key where {@code key} exists in the table
+     */
+    int get(Object key);
```

Comment:
Yes, given that we have no plans to enhance our hash tables to support more than `Integer.MAX_VALUE` rows in the forseeable future. If and when we change that, we can change this interface accordingly, since it is quite internal in usage. If we don't have a constrained number of bits, at least, I'll need to change up some of the rollup node ID code at that time.